drivers/spi/spi_context.h

```diff
@@ -112,13 +112,27 @@ static inline void spi_context_release(struct spi_context *ctx, int status)
 static inline int spi_context_wait_for_completion(struct spi_context *ctx)
 {
 	int status = 0;
+	uint64_t ms;
+
+	ms = MAX(ctx->tx_len, ctx->rx_len) * 8 * 1000 / ctx->config->frequency;
+	ms += CONFIG_SPI_COMPLETION_TIMEOUT_TOLERANCE;
+
+	if (ms > UINT_MAX)
+		ms = UINT_MAX;
+
 #ifdef CONFIG_SPI_ASYNC
 	if (!ctx->asynchronous) {
-		k_sem_take(&ctx->sync, K_FOREVER);
+		if (k_sem_take(&ctx->sync, K_MSEC(ms))) {
+			LOG_ERR("Timeout waiting for transfer complete");
+			return -ETIMEDOUT;
+		}
```

Comment:
It would be cleaner if we passed on the value returned by `k_sem_take` and not fixed `-ETIMEDOUT`.

Reply:
Actually, I'm not sure about this anymore. Maybe `-ETIMEDOUT` is a better choice.